pycontrails/datalib/goes.py

```diff
@@ -0,0 +1,730 @@
+
+        lpaths = self._lpaths(time)
+
+        channels_needed = set()
+        for c, lpath in lpaths.items():
+            if not self.cachestore.exists(lpath):
+                channels_needed.add(c)
+
+        if channels_needed:
+            rpaths = self.gcs_goes_path(time, channels_needed)
+            for rpath in rpaths:
+                channel = _extract_channel_from_rpath(rpath)
+                lpath = lpaths[channel]
+                self.fs.get(rpath, lpath)
+
+        # Deal with the different spatial resolutions
+        lpath02 = lpaths.pop("C02", None)
+        kwargs = {
+            "paths": lpaths.values(),
+            "concat_dim": "band",
+            "combine": "nested",
+            "data_vars": ["CMI"],
+            "compat": "override",
+            "coords": "minimal",
+        }
+        if lpath02:
+            ds1 = xr.open_mfdataset(**kwargs)
+            ds2 = xr.open_dataset(lpath02)
+
+            # Average the C02 data to the C01 resolution
+            ds2 = ds2.coarsen(x=2, y=2, boundary="exact").mean()
+
+            # Gut check
+            np.testing.assert_allclose(ds1["x"], ds2["x"], rtol=0.0005)
+            np.testing.assert_allclose(ds1["y"], ds2["y"], rtol=0.0005)
+
+            # Assign the C01 data to the C02 data
+            ds2["x"] = ds1["x"]
+            ds2["y"] = ds1["y"]
+
+            # Finally, combine the datasets
+            ds = xr.concat([ds1, ds2], dim="band")
+        else:
+            ds = xr.open_mfdataset(**kwargs)
+
+        da = ds["CMI"]
+        da = da.swap_dims({"band": "band_id"}).sortby("band_id")
+
+        # Attach some useful attrs -- only using goes_imager_projection currently
+        da.attrs["goes_imager_projection"] = ds.goes_imager_projection.attrs
+        da.attrs["geospatial_lat_lon_extent"] = ds.geospatial_lat_lon_extent.attrs
+
+        return da
+
+    def _get_without_cache(self, time: datetime.datetime) -> xr.DataArray:
+        """Download the GOES data into memory at the given time."""
+        rpaths = self.gcs_goes_path(time)
+
+        # Load into memory
+        data = self.fs.cat(rpaths)
+
+        if isinstance(data, dict):
+            da_list = []
+            for init_bytes in data.values():
+                ds = xr.open_dataset(io.BytesIO(init_bytes), engine="h5netcdf")
+
+                da = ds["CMI"]
+                da = da.expand_dims(band_id=ds["band_id"].values)
+                da_list.append(da)
+
+            da = xr.concat(da_list, dim="band_id")
+        else:
+            ds = xr.open_dataset(io.BytesIO(data), engine="h5netcdf")
+            da = ds["CMI"]
+            da = da.expand_dims(band_id=ds["band_id"].values)
+
+        # Attach some useful attrs -- only using goes_imager_projection currently
+        da.attrs["goes_imager_projection"] = ds.goes_imager_projection.attrs
+        da.attrs["geospatial_lat_lon_extent"] = ds.geospatial_lat_lon_extent.attrs
+
+        return da
+
+
+def extract_goes_viz_artifacts(
```

Comment:
I'm not keen on this name - what if it was `extract_goes_visualization` or `extract_goes_image`?

Reply:
481d223b7a633578209b1f10fd5f3d5d9c8fc8b1